worker/provisioner/provisioner_task.go

```diff
@@ -689,101 +723,362 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+	var dgAvailabilityZoneMachines []*AvailabilityZoneMachine
+	dgSet := set.NewStrings(machineIds...)
+	for _, azm := range task.availabilityZoneMachines {
+		dgAvailabilityZoneMachines = append(dgAvailabilityZoneMachines, &AvailabilityZoneMachine{
+			azm.ZoneName,
+			azm.MachineIds.Intersection(dgSet),
+			azm.FailedMachineIds,
+		})
+	}
+	return dgAvailabilityZoneMachines
+}
 
-		var arch string
-		if pInfo.Constraints.Arch != nil {
-			arch = *pInfo.Constraints.Arch
-		}
+// machineAvailabilityZoneDistribution returns a suggested availability zone
+// for the specified machine to start in.  If the current provider does not
+// implement availability zones, "" will be returned. Machines are spread across
+// availability zones based on lowest population of the "available" zones.
+// Machines in the same DistributionGroup are placed in different zones, spread
+// across availability zones based on lowest population of machines in that
+// DistributionGroup.
+func (task *provisionerTask) machineAvailabilityZoneDistribution(machine *apiprovisioner.Machine, distributionGroupMachineIds []string) string {
+	task.azMachinesMutex.Lock()
+	defer task.azMachinesMutex.Unlock()
+
+	if len(task.availabilityZoneMachines) == 0 {
+		return ""
+	}
 
-		v, err := m.ModelAgentVersion()
-		if err != nil {
-			return errors.Trace(err)
+	var machineZone string
+	// assign an initial az to a machine based on lowest population.
+	// if the machine has a distribution group, assign based on lowest
+	// az population of the distribution group machine.
+	if len(distributionGroupMachineIds) > 0 {
+		dgZoneMap := task.populateDistributionGroupZoneMap(distributionGroupMachineIds)
+		sort.Sort(ByPopulationThenName(dgZoneMap))
+
+		for _, dgZoneMachines := range dgZoneMap {
+			if !dgZoneMachines.FailedMachineIds.Contains(machine.Id()) {
+				machineZone = dgZoneMachines.ZoneName
+				for _, azm := range task.availabilityZoneMachines {
+					if azm.ZoneName == dgZoneMachines.ZoneName {
+						azm.MachineIds.Add(machine.Id())
+						break
+					}
+				}
+				break
+			}
 		}
-		possibleTools, err := task.toolsFinder.FindTools(
-			*v,
-			pInfo.Series,
-			arch,
-		)
-		if err != nil {
-			return task.setErrorStatus("cannot find tools for machine %q: %v", m, err)
+	} else {
+		sort.Sort(ByPopulationThenName(task.availabilityZoneMachines))
+		for _, zoneMachines := range task.availabilityZoneMachines {
+			if !zoneMachines.FailedMachineIds.Contains(machine.Id()) {
+				machineZone = zoneMachines.ZoneName
+				zoneMachines.MachineIds.Add(machine.Id())
+				break
+			}
 		}
+	}
+	return machineZone
+}
 
-		startInstanceParams, err := constructStartInstanceParams(
-			task.controllerUUID,
-			m,
-			instanceCfg,
-			pInfo,
-			possibleTools,
-		)
-		if err != nil {
-			return task.setErrorStatus("cannot construct params for machine %q: %v", m, err)
+type ByPopulationThenName []*AvailabilityZoneMachine
```

Comment:
don't think this needs to be exported

Reply:
It's used in the tests, is there a way to have export_test allow for that?